vm/src/vm.rs

```diff
@@ -941,21 +941,45 @@ impl VirtualMachine {
         }
     }
 
+    pub fn _isinstance(&self, obj: &PyObjectRef, cls: &PyObjectRef) -> PyResult<bool> {
+        if let Ok(typ) = PyTypeRef::try_from_object(self, cls.clone()) {
+            Ok(obj.isinstance(&typ))
+        } else {
+            Err(self.new_type_error(format!(
+                "isinstance() arg 2 must be a type or tuple of types, not {}",
+                cls.class()
+            )))
+        }
+    }
+
     /// Determines if `obj` is an instance of `cls`, either directly, indirectly or virtually via
     /// the __instancecheck__ magic method.
-    pub fn isinstance(&self, obj: &PyObjectRef, cls: &PyTypeRef) -> PyResult<bool> {
+    pub fn isinstance(&self, obj: &PyObjectRef, cls: &PyObjectRef) -> PyResult<bool> {
         // cpython first does an exact check on the type, although documentation doesn't state that
-        // https://github.com/python/cpython/blob/a24107b04c1277e3c1105f98aff5bfa3a98b33a0/Objects/abstract.c#L2408
+        // https://github.com/python/cpytho(n/blob/a24107b04c1277e3c1105f98aff5bfa3a98b33a0/Objects/abstract.c#L2408
         if obj.class().is(cls) {
-            Ok(true)
-        } else {
-            let ret = self.call_special_method(
-                cls.as_object().clone(),
-                "__instancecheck__",
-                (obj.clone(),),
-            )?;
-            pybool::boolval(self, ret)
+            return Ok(true);
         }
+
+        if self.ctx.types.type_type.is(&cls.clone_class()) {
```

Comment:
```suggestion
        if self.ctx.types.type_type.is(&cls.clone_class()) 
```

Reply:
Oh you are right. no need to use `clone_class`. Thanks!

Reply:
```suggestion
        if cls.lease_class().is(self.ctx.types.type_type) {
```

Reply:
I pasted the wrong thing :laughing: I wanted to suggest `cls.class().is(self.ctx.types.type_type)`. Any difference between that and `lease_class` @youknowone ?

Reply:
I think his intend was your suggestion code :) I apply your suggestion and add expected failure to
failed test in `test_isinstance.py`.

Reply:
@DimitrisJim `lease_class` doesn't exist any more thanks to a089c9b. I think @youknowone might have slipped up.